h2o-core/src/main/java/hex/grid/HyperSpaceWalker.java

```diff
@@ -489,44 +505,82 @@ public MP nextModelParameters(Model previousModel) {
-              // ToDo: model seed setting will be different for parallel model building.
-              // ToDo: This implementation only works for sequential model building.
-              if (_set_model_seed_from_search_seed) {
-                // set model seed = search_criteria.seed+(0, 1, 2,..., model number)
-                params._seed = _search_criteria.seed() + (model_number++);
-              }
+             
+            if (_filterFunctions == null || hyperParamsAreNotSkipped(hypers, _filterFunctions)) {
+
+              _numberOfNonSkippedPermutations++;
+              // Get clone of parameters
+              MP commonModelParams = (MP) _params.clone();
+              // Fill model parameters
+              MP params = getModelParams(commonModelParams, hypers);
+
+              // add max_runtime_secs in search criteria into params if applicable
+              if (_search_criteria != null && _search_criteria.strategy() == HyperSpaceSearchCriteria.Strategy.RandomDiscrete) {
+                // ToDo: model seed setting will be different for parallel model building.
+                // ToDo: This implementation only works for sequential model building.
+                if (_set_model_seed_from_search_seed) {
+                  // set model seed = search_criteria.seed+(0, 1, 2,..., model number)
+                  params._seed = _search_criteria.seed() + (model_number++);
+                }
 
-              // set max_runtime_secs
-              double timeleft = this.time_remaining_secs();
-              if (timeleft > 0)  {
-                if (params._max_runtime_secs > 0) {
-                  params._max_runtime_secs = min(params._max_runtime_secs, timeleft);
-                } else {
-                  params._max_runtime_secs = timeleft;
+                // set max_runtime_secs
+                double timeleft = this.time_remaining_secs();
+                if (timeleft > 0) {
+                  if (params._max_runtime_secs > 0) {
+                    params._max_runtime_secs = min(params._max_runtime_secs, timeleft);
+                  } else {
+                    params._max_runtime_secs = timeleft;
+                  }
                 }
               }
+              return params;
+            } else {
+              
+              int nextPermutationNum = _currentPermutationNum + 1;
+              if(hasNext(previousModel) && (max_models() == 0 || max_models() > 0 && nextPermutationNum < max_models())) {
+                return nextModelParameters(previousModel);
+              } else {
+                return null;
+              }
             }
-            return params;
           } else {
             throw new NoSuchElementException("No more elements to explore in hyper-space!");
           }
         }
+        
+        private boolean hyperParamsAreNotSkipped(Object[] hypers, ArrayList<Function<Map<String, Object>, Boolean>> filterFunctions) {
```

Comment:
as soon as I see a variable/method with a negation, I know that my head is gonna hurt later due to double negations... is it necessary?

Reply:
@seb-h2o I understand your concern,  but when I see something is called `doSomething` it means I would expect method to do THIS and it is not necessary should return boolean to report its status.
Anyway,  all those negations and amount of efforts to understand them depend on the context. I could have named it `areHyperParamsSkipped` to avoid negation but I would have to put `!` sign in front of it - and this also complicates reading. So my point is - with boolean logic it is almost unavoidable that it will make somebody’s head to hurt.

Reply:
maybe boolean logic can hurt, but it always hurts stronger with double negations